src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/GenericVirtualMethodImplNode.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Collections.Generic;
+
+using ILCompiler.DependencyAnalysisFramework;
+
+using Internal.TypeSystem;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    /// <summary>
+    /// Represents a use of a generic virtual method body implementation.
+    /// </summary>
+    public class GenericVirtualMethodImplNode : DependencyNodeCore<NodeFactory>
+    {
+        private const int UniversalCanonGVMDepthHeuristic_CanonDepth = 2;
+        private readonly MethodDesc _method;
+
+        public GenericVirtualMethodImplNode(MethodDesc method)
+        {
+            Debug.Assert(method.GetCanonMethodTarget(CanonicalFormKind.Specific) == method);
+            Debug.Assert(method.HasInstantiation);
+
+            // This is either a generic virtual method or a MethodImpl for a static interface method.
+            // We can't test for static MethodImpl so at least sanity check it's static and noninterface.
+            Debug.Assert(method.IsVirtual || (method.Signature.IsStatic && !method.OwningType.IsInterface));
+
+            _method = method;
+        }
+
+        public override bool HasConditionalStaticDependencies => false;
+        public override bool InterestingForDynamicDependencyAnalysis => false;
+        public override bool StaticDependenciesAreComputed => true;
+        protected override string GetName(NodeFactory factory) => "__GVMImplNode_" + factory.NameMangler.GetMangledMethodName(_method);
+
+        public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFactory factory)
+        {
+            DependencyList dependencies = null;
+
+            factory.MetadataManager.GetDependenciesDueToVirtualMethodReflectability(ref dependencies, factory, _method);
+
+            bool validInstantiation =
+                _method.IsSharedByGenericInstantiations || (      // Non-exact methods are always valid instantiations (always pass constraints check)
+                    _method.Instantiation.CheckValidInstantiationArguments() &&
+                    _method.OwningType.Instantiation.CheckValidInstantiationArguments() &&
+                    _method.CheckConstraints());
+
+            if (validInstantiation)
```

Comment:
By "valid" here, do we mean that the code is not broken? That the compiler hasn't written incorrect IL?

Reply:
This is a cargo cult I brought over from the previous spot. I don't know when it kicks in. It goes back all the way to https://github.com/dotnet/corert/pull/2521.